File: Golang/src/pkg/util/passwordfile.go
package util

import (
	"errors"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const original = "0123456789"

type FilePassword struct {
	fileName string
	password string
}

// NewFilePassword 创建FilePassword实例，检查密码是否包含重复字符
func NewFilePassword(filename, password string) *FilePassword {
	fp := &FilePassword{
		fileName: filename,
		password: validatePassword(password),
	}
	return fp
}

// NewFilePasswordWithDefault 创建FilePassword实例，如果文件不存在则写入默认值
func NewFilePasswordWithDefault(filename, password string, defaultValue int) *FilePassword {
	fp := NewFilePassword(filename, password)
	
	// 检查文件是否存在
	if _, err := os.Stat(filename); os.IsNotExist(err) {
		fp.Write(defaultValue)
	}
	return fp
}

// validatePassword 验证密码是否包含重复字符
func validatePassword(password string) string {
	charSet := make(map[rune]bool)
	for _, c := range password {
		if charSet[c] {
			return "1324096857" // 使用默认密码
		}
		charSet[c] = true
	}
	return password
}

// translateChar 字符转换辅助函数
func translateChar(c byte, from, to string) (byte, error) {
	index := strings.IndexByte(from, c)
	if index == -1 {
		return 0, errors.New("character not found in mapping table")
	}
	return to[index], nil
}

// ReadPassword 读取文件中的密码
func (fp *FilePassword) ReadPassword() (string, error) {
	// 检查文件是否存在，不存在则创建
	if _, err := os.Stat(fp.fileName); os.IsNotExist(err) {
		file, err := os.Create(fp.fileName)
		if err != nil {
			return "", fmt.Errorf("failed to create file: %s", fp.fileName)
		}
		// 写入密码的第一个字符
		file.WriteString(string(fp.password[0]))
		file.Close()
	}

	// 读取文件内容
	content, err := os.ReadFile(fp.fileName)
	if err != nil {
		return "", fmt.Errorf("failed to open file: %s", fp.fileName)
	}

	return string(content), nil
}

// ReadReal 将文件中的加密密码转换为真实数字
func (fp *FilePassword) ReadReal() (string, error) {
	encrypted, err := fp.ReadPassword()
	if err != nil {
		return "", err
	}

	result := ""
	for _, c := range encrypted {
		translated, err := translateChar(byte(c), fp.password, original)
		if err != nil {
			return "", errors.New("error in character translation")
		}
		result += string(translated)
	}
	return result, nil
}

// Write 将数字加密后写入文件
func (fp *FilePassword) Write(newNum int) error {
	numStr := strconv.Itoa(newNum)
	result := ""
	for _, c := range numStr {
		translated, err := translateChar(byte(c), original, fp.password)
		if err != nil {
			return errors.New("invalid number format")
		}
		result += string(translated)
	}

	file, err := os.Create(fp.fileName)
	if err != nil {
		return fmt.Errorf("failed to open file for writing: %s", fp.fileName)
	}
	defer file.Close()

	_, err = file.WriteString(result)
	return err
}

// Add 读取文件中的数字，加上指定值后再加密写入文件
func (fp *FilePassword) Add(num int) error {
	realStr, err := fp.ReadReal()
	if err != nil {
		return errors.New("error reading real value")
	}

	currentNum, err := strconv.Atoi(realStr)
	if err != nil {
		return errors.New("invalid number in file")
	}

	newNum := currentNum + num
	return fp.Write(newNum)
}

// ReadStr 读取存储数字返回string
func (fp *FilePassword) ReadStr() (string, error) {
	realStr, err := fp.ReadReal()
	if err != nil {
		return "", err
	}
	
	// 转换为数字再转回字符串，确保格式正确
	if num, err := strconv.Atoi(realStr); err == nil {
		return strconv.Itoa(num), nil
	}
	return "", errors.New("invalid number format")
}

// ReadInt 读取存储数字返回int
func (fp *FilePassword) ReadInt() (int, error) {
	realStr, err := fp.ReadReal()
	if err != nil {
		return 0, err
	}
	return strconv.Atoi(realStr)
}

// AddNum 增加存储数字
func (fp *FilePassword) AddNum(a int) error {
	if a < 0 {
		return nil // 负数不操作
	}
	return fp.Add(a)
}

// MinusNum 减小存储数字，不检查是否为负数
func (fp *FilePassword) MinusNum(a int) error {
	if a < 0 {
		return nil // 负数不操作
	}
	return fp.Add(-a)
}

// MinusNumIf 减小存储数字，若不足则减至0
func (fp *FilePassword) MinusNumIf(a int) error {
	if a < 0 {
		return nil
	}

	current, err := fp.ReadInt()
	if err != nil {
		return err
	}

	if current >= a {
		return fp.MinusNum(a)
	} else {
		return fp.MinusNum(current)
	}
}

// High 判断是否大于等于指定数字
func (fp *FilePassword) High(a int) (bool, error) {
	current, err := fp.ReadInt()
	if err != nil {
		return false, err
	}
	return current >= a, nil
}

// CanMinus 判断是否大于等于指定数字，若满足则减去，不满足则不变
func (fp *FilePassword) CanMinus(a int) (bool, error) {
	if a < 0 {
		return false, nil
	}

	isHigh, err := fp.High(a)
	if err != nil {
		return false, err
	}

	if isHigh {
		err = fp.MinusNum(a)
		if err != nil {
			return false, err
		}
		return true, nil
	}
	return false, nil
}

func (fp *FilePassword) ReadStrSafe() string {
	str, _ := fp.ReadStr()
	return str
}
func (fp *FilePassword) ReadIntSafe() int {
	inte, _ := fp.ReadInt()
	return inte
}

File: src-tauri/src/util/passwordfile.rs
//! 兼容的密码存取实现

use std::collections::HashSet;
use std::fs::{self, File};
use std::io::Write;
use std::path::Path;

const ORIGINAL: &str = "0123456789";

pub struct FilePassword {
    file_name: String,
    password: String,
}

impl FilePassword {
    /// 创建 FilePassword 实例，检查密码是否包含重复字符
    pub fn new(filename: &str, password: &str) -> Self {
        let validated_password = Self::validate_password(password);
        FilePassword {
            file_name: filename.to_string(),
            password: validated_password,
        }
    }

    /// 创建 FilePassword 实例，如果文件不存在则写入默认值
    pub fn new_with_default(
        filename: &str, 
        password: &str, 
        default_value: i32
    ) -> Result<Self, crate::error::SpError> {
        let fp = Self::new(filename, password);
        
        // 检查文件是否存在
        if !Path::new(filename).exists() {
            fp.write(default_value)?;
        }
        Ok(fp)
    }

    /// 验证密码是否包含重复字符
    fn validate_password(password: &str) -> String {
        let mut char_set = HashSet::new();
        for c in password.chars() {
            if char_set.contains(&c) {
                return "1324096857".to_string(); // 使用默认密码
            }
            char_set.insert(c);
        }
        password.to_string()
    }

    /// 字符转换辅助函数
    fn translate_char(c: char, from: &str, to: &str) -> Result<char, crate::error::SpError> {
        from.find(c)
            .and_then(|idx| to.chars().nth(idx))
            .ok_or_else(|| {
                crate::error::SpError::ErrorTranslation(
                    format!("character '{}' not found in mapping table", c)
                )
            })
    }

    /// 读取文件中的密码
    pub fn read_password(&self) -> Result<String, crate::error::SpError> {
        // 检查文件是否存在，不存在则创建
        if !Path::new(&self.file_name).exists() {
            let mut file = File::create(&self.file_name)
                .map_err(|e| crate::error::SpError::ErrorIO(format!("failed to create file: {}", e)))?;
            
            // 写入密码的第一个字符
            if let Some(first_char) = self.password.chars().next() {
                write!(file, "{}", first_char)
                    .map_err(|e| crate::error::SpError::ErrorIO(format!("failed to write to file: {}", e)))?;
            }
        }

        // 读取文件内容
        let content = fs::read_to_string(&self.file_name)
            .map_err(|e| crate::error::SpError::ErrorIO(format!("failed to read file: {}", e)))?;
        
        Ok(content.trim().to_string())
    }

    /// 将文件中的加密密码转换为真实数字
    pub fn read_real(&self) -> Result<String, crate::error::SpError> {
        let encrypted = self.read_password()?;
        
        let mut result = String::new();
        for c in encrypted.chars() {
            let translated = Self::translate_char(c, &self.password, ORIGINAL)?;
            result.push(translated);
        }
        Ok(result)
    }

    /// 将数字加密后写入文件
    pub fn write(&self, new_num: i32) -> Result<(), crate::error::SpError> {
        let num_str = new_num.to_string();
        let mut result = String::new();
        
        for c in num_str.chars() {
            match Self::translate_char(c, ORIGINAL, &self.password) {
                Ok(translated) => result.push(translated),
                Err(e) => return Err(crate::error::SpError::ErrorIO(
                    format!("invalid number format during translation: {}", e)
                )),
            }
        }

        let mut file = File::create(&self.file_name)
            .map_err(|e| crate::error::SpError::ErrorIO(format!("failed to create file for writing: {}", e)))?;
        
        file.write_all(result.as_bytes())
            .map_err(|e| crate::error::SpError::ErrorIO(format!("failed to write to file: {}", e)))?;
        
        Ok(())
    }

    /// 读取文件中的数字，加上指定值后再加密写入文件
    pub fn add(&self, num: i32) -> Result<(), crate::error::SpError> {
        let real_str = self.read_real()?;
        
        let current_num: i32 = real_str
            .parse()
            .map_err(|e| crate::error::SpError::ErrorIO(
                format!("failed to parse number from file: {}", e)
            ))?;
        
        let new_num = current_num + num;
        self.write(new_num)?;
        Ok(())
    }

    /// 读取存储数字返回 string
    pub fn read_str(&self) -> Result<String, crate::error::SpError> {
        let real_str = self.read_real()?;
        
        // 转换为数字再转回字符串，确保格式正确
        match real_str.parse::<i32>() {
            Ok(num) => Ok(num.to_string()),
            Err(e) => Err(crate::error::SpError::ErrorIO(
                format!("invalid number format: {}", e)
            )),
        }
    }

    /// 读取存储数字返回 i32
    pub fn read_int(&self) -> Result<i32, crate::error::SpError> {
        let real_str = self.read_real()?;
        
        real_str
            .parse()
            .map_err(|e| crate::error::SpError::ErrorIO(
                format!("failed to parse integer: {}", e)
            ))
    }

    /// 增加存储数字
    pub fn add_num(&self, a: i32) -> Result<(), crate::error::SpError> {
        if a < 0 {
            return Ok(()); // 负数不操作
        }
        self.add(a)
    }

    /// 减小存储数字，不检查是否为负数
    pub fn minus_num(&self, a: i32) -> Result<(), crate::error::SpError> {
        if a < 0 {
            return Ok(()); // 负数不操作
        }
        self.add(-a)
    }

    /// 减小存储数字，若不足则减至0
    pub fn minus_num_if(&self, a: i32) -> Result<(), crate::error::SpError> {
        if a < 0 {
            return Ok(());
        }

        let current = self.read_int()?;

        if current >= a {
            self.minus_num(a)
        } else {
            self.minus_num(current)
        }
    }

    /// 判断是否大于等于指定数字
    pub fn high(&self, a: i32) -> Result<bool, crate::error::SpError> {
        let current = self.read_int()?;
        Ok(current >= a)
    }

    /// 判断是否大于等于指定数字，若满足则减去，不满足则不变
    pub fn can_minus(&self, a: i32) -> Result<bool, crate::error::SpError> {
        if a < 0 {
            return Ok(false);
        }

        let is_high = self.high(a)?;

        if is_high {
            self.minus_num(a)?;
            Ok(true)
        } else {
            Ok(false)
        }
    }

    /// 安全的读取字符串版本（忽略错误）
    pub fn read_str_safe(&self) -> String {
        self.read_str().unwrap_or_default()
    }

    /// 安全的读取整数版本（忽略错误）
    pub fn read_int_safe(&self) -> i32 {
        self.read_int().unwrap_or(0)
    }

    /// 安全的 CanMinus 版本（忽略错误）
    pub fn can_minus_safe(&self, an: i32) -> bool {
        self.can_minus(an).unwrap_or(false)
    }
}